appcelerator/titanium_mobile
# feat(android): Added Ti.Platform.canOpenURL() support

**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-14460

**Summary:**
- [TIMOB-14460](https://jira.appcelerator.org/browse/TIMOB-14460) Added `Ti.Platform.canOpenURL()` support. _(Parity with iOS and Windows.)_
- Added ability to open app's sandboxed files via `openURL()` and `canOpenURL()`.
  * iOS supports something like this via [Ti.UI.iOS.DocumentViewer](https://docs.appcelerator.com/platform/latest/#!/api/Titanium.UI.iOS.DocumentViewer) which displays file in-app.
  * This provides a similar equivalent on Android. (We should do the same for `startActivity()` method in the future.)
- Updated `TiFileProvider.java` to provide external access to embedded files within the APK's "assets" and "res" folders.
- Updated `TiMimeTypeHelper.java` to fetch mime type from image, audio, and video files if they don't have a file extension.
  * Requires scanning the binary. Needed for APK "res" files not under "res/raw".
- Unit test already written in mocha suite. Need to remove `it.androidMissing` when merged.

**Test:**
1. Create a new Titanium Classic "Default Project".
2. Copy the below code to the proejct's "app.js" file.
3. Build and run on Android.
4. Notice that a table view of tests are displayed. Tap on the "canOpenURL" and "openURL" buttons for each row and verify it matches the results below.

| URL | canOpenURL() | openURL() |
| --- | ------------ | --------- |
| `"http://www.appcelerator.com` | `true` | Opens Appcelerator website in web browser app. |
| `https://www.appcelerator.com` | `true` | Opens Appcelerator website in web browser app. |
| `mailto:user@domain.com` | `true` | Opens new draft in e-mail app with "user@domain.com" assigned to "To" field. |
| `mailto:user@domain.com?cc=johndoe@domain.com ...` | `true` | Opens new draft in e-mail app. "TO" field set to "user@domain.com". "CC" field set to "johndoe@domain.com" and "janedoe@domain.com". "BCC" field set to "bsmith@domain.com". "Subject" field set to "The Subject". "Body" field set to "This is a test. Hello World!". |
| `tel:123-456-7890` | `true` on phone. `false` on tablet. | Opens phone app with phone number set to "(123) 456-7890". |
| `geo:37.774,-122.431?z=12` | `true` | Displays San Francisco in maps app. |
| `maps:?address=London,UK` | `false` | Won't open. (Only supported on iOS.) |
| `badscheme://i.dont.exist` | `false` | Won't open because URL scheme doesn't exist. |
| `assets/images/tab1.png` | `true` | Displays appc logo in photo app. (Won't work in Android's old "gallery" app.) |
| `/assets/images/tab1.png` | `true` | Displays appc logo in photo app. (Won't work in Android's old "gallery" app.) |
| `././assets/images/.././images/tab1.png` | `true` | Displays appc logo in photo app. (Won't work in Android's old "gallery" app.) |
| `file:///android_asset/ ... /tab1@2x.png` | `true` | Displays **large** appc logo in photo app. (Won't work in Android's old "gallery" app.) |
| `appdata-private://Copy of tab1@2x.png` | `true` | Displays **large** appc logo in photo app. (Won't work in Android's old "gallery" app.) |
| `file:///data/ ... /Copy of tab1@2x.png` | `true` | Displays **large** appc logo in photo app. (Won't work in Android's old "gallery" app.) |
| `content:// ... /tab1.png` | `true` | Displays appc logo in photo app. (Won't work in Android's old "gallery" app.) |
| `app://assets/images/tab1.png` | `true` | Displays appc logo in photo app. (Won't work in Android's old "gallery" app.) |
| `file:///android_asset/ ... /tab1.png` | `true` | Displays appc logo in photo app. (Won't work in Android's old "gallery" app.) |
| `android.resource:// ... /drawable/appicon` | `true` | Displays app's large icon in photo app. (Won't work in Android's old "gallery" app.) |

app.js
```javascript
function createTableRowForUrl(url) {
	var row = Ti.UI.createTableViewRow({
		layout: "horizontal",
	});
	row.add(Ti.UI.createLabel({
		text: url,
		top: "10dp",
		left: "10dp",
		right: "10dp",
		width: Ti.UI.FILL,
		height: Ti.UI.SIZE,
	}));
	var canOpenButton = Ti.UI.createButton({
		title: "canOpenURL()",
		top: "10dp",
		bottom: "10dp",
		left: "10dp",
		width: Ti.UI.SIZE,
		height: Ti.UI.SIZE,
	});
	canOpenButton.addEventListener("click", function() {
		var canOpen = Ti.Platform.canOpenURL(url);
		Ti.API.info("@@@ canOpenURL() returned '" + canOpen + "' for URL: " + url);
		alert("canOpenURL() returned:\n" + canOpen);
	});
	row.add(canOpenButton);
	var openButton = Ti.UI.createButton({
		title: "openURL()",
		top: "10dp",
		bottom: "10dp",
		left: "10dp",
		width: Ti.UI.SIZE,
		height: Ti.UI.SIZE,
	});
	openButton.addEventListener("click", function() {
		var wasOpened = Ti.Platform.openURL(url);
		Ti.API.info("@@@ openURL() returned '" + wasOpened + "' for URL: " + url);
	});
	row.add(openButton);
	return row;
}

var tableRows = [];
tableRows.push(createTableRowForUrl("http://www.appcelerator.com"));
tableRows.push(createTableRowForUrl("https://www.appcelerator.com"));
tableRows.push(createTableRowForUrl("mailto:user@domain.com"));
tableRows.push(createTableRowForUrl("mailto:user@domain.com?cc=johndoe@domain.com,janedoe@domain.com&bcc=bsmith@domain.com&subject=The%20Subject&body=This%20is%20a%20test.%0AHello%20World!"));
tableRows.push(createTableRowForUrl("tel:123-456-7890"));
tableRows.push(createTableRowForUrl("geo:37.774,-122.431?z=12"));  // San Francisco
tableRows.push(createTableRowForUrl("maps:?address=London,UK"));
tableRows.push(createTableRowForUrl("badscheme://i.dont.exist"));
if (Ti.Android) {
	var sourceFile = Ti.Filesystem.getFile(Ti.Filesystem.resourcesDirectory, "assets/images/tab1@2x.png");
	var externalFilePath = Ti.Filesystem.applicationDataDirectory + "Copy of " + sourceFile.name;
	sourceFile.copy(externalFilePath);
	tableRows.push(createTableRowForUrl("assets/images/tab1.png"));
	tableRows.push(createTableRowForUrl("/assets/images/tab1.png"));
	tableRows.push(createTableRowForUrl("././assets/images/.././images/tab1.png"));
	tableRows.push(createTableRowForUrl(sourceFile.nativePath));
	tableRows.push(createTableRowForUrl(externalFilePath));
	tableRows.push(createTableRowForUrl(Ti.Filesystem.getFile(externalFilePath).nativePath));
	tableRows.push(createTableRowForUrl("content://" + Ti.App.id + ".tifileprovider/assets/Resources/assets/images/tab1.png"));
	tableRows.push(createTableRowForUrl(Ti.Filesystem.resourcesDirectory + "assets/images/tab1.png"));
	tableRows.push(createTableRowForUrl("file:///android_asset/Resources/assets/images/tab1.png"));
	tableRows.push(createTableRowForUrl("android.resource://" + Ti.App.id + "/drawable/appicon"));
}

var window = Ti.UI.createWindow({
	extendSafeArea: false,
	fullscreen: true,
	theme: "Theme.AppCompat.NoTitleBar",
});
var tableView = Ti.UI.createTableView({
	data: tableRows,
});
window.add(tableView);
window.open();
```
